### PR TITLE
chore(appium-service): filter out `Debugger attached` as errorMessage

### PR DESCRIPTION
**Pull request by Delta456**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

Fixes https://github.com/webdriverio/webdriverio/issues/12227 by filtering out `Debugger attached` in log error as said by @christian-bromann in https://github.com/webdriverio/webdriverio/issues/12227#issuecomment-1940407447

With the fix, `ERROR: Debugger attached` is now gone. 

image

### CODE REVIEW

**Comment by Delta456** [comment]
I am not sure how will I test this change and if this needs to be backported to v8 as well

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Can we add a unit test for this?

**Comment by Delta456** [comment]
I am not sure how we will test these changes Do you know how :thinking:

**Comment by christian-bromann** [comment]
I recommend take a look at [these unit tests](https://github.com/webdriverio/webdriverio/blob/bd45bf9b879b4947d2adfa0197ee7b8e2d027774/packages/wdio-appium-service/tests/launcher.test.ts#L742-L774) and add one that mocks the child process in a similar way but triggers the stderr event to be called so you can test your callback.

**Comment by christian-bromann** [comment]
@Delta456 if you have trouble with the unit test, feel free to schedule a 1:1 session through our [Open Office Hours](https://webdriver.io/community/openofficehours), thanks.

**Comment by Delta456** [comment]
> @Delta456 if you have trouble with the unit test, feel free to schedule a 1:1 session through our [Open Office Hours](https://webdriver.io/community/openofficehours), thanks.

Hi @christian-bromann, I have scheduled a session with you. Hope to see you there

**Comment by Delta456** [comment]
The test is now added. Thanks to the appium team!

File: packages/wdio-appium-service/tests/launcher.test.ts [hunk 1]
```diff
@@ -771,6 +774,22 @@ describe('Appium launcher', () => {
                 10000
             )).resolves.toEqual(expect.objectContaining({ spawnargs: expect.arrayContaining(['-e', expect.any(String)]) }))
         })
+
+        test('filter out "Debugger attached" message as an error', async () => {
+            const origSpawn = await vi.importActual<typeof cp>('node:child_process').then((m) => m.spawn)
+            vi.mocked(spawn).mockImplementationOnce(origSpawn)
```
Comment (christian-bromann) [inline packages/wdio-appium-service/tests/launcher.test.ts:780]: I don't think we want to actually start Appium in the unit test. Instead we should mock the `spawn` call and return a mock that allows us to execute the error handler and verify if that error handler works correctly, here is an example:

```ts
const eventListener = { on: vi.fn(), off: vi.fn(), once: vi.fn() }
vi.mocked(spawn).mockReturnValue({
    ...eventListener,
    stdout: { ...eventListener },
    stderr: { ...eventListener },
})

// call function under  test
...

// get error handler
expect(eventListener.on).toBeCalledWith('data', expect.any(Function))
const errorHandler = spawn.mock.calls[0][1].on

// actually call error handler
errorHandler(Buffer.from('some error message'))
expect(mockLogError.error).toBeCalledWith('some error message')
mockLogError.error.mockClear()

errorHandler(Buffer.from('Debugger attached.'))
expect(mockLogError.error).not.toBeCalled()
```

**Review by christian-bromann — CHANGES_REQUESTED** [review]
One suggestion on the unit test.

File: packages/wdio-appium-service/tests/launcher.test.ts (hunk 1 above)
Comment (Delta456) [reply packages/wdio-appium-service/tests/launcher.test.ts:780]: I think both serve the same purpose and why isn't launching appium a good idea? I would want an explanation to understand 😅

Comment (SrinivasanTarget) [reply packages/wdio-appium-service/tests/launcher.test.ts:780]: @Delta456 In your unit tests, yes you are starting the node process which simulates the appium behavior. Whereas in @christian-bromann suggestion, he don't start node process and mocks spawn function to return an object with event listeners. This approach directly triggers error handler and test the behavior.

+1 to Christian's suggestion coz its faster and test the error handler functionality directly.

Comment (christian-bromann) [reply packages/wdio-appium-service/tests/launcher.test.ts:780]: @Delta456 unit tests should only focus on testing code and should not have any dependencies to 3rd party resources like an Appium server.

Comment (Delta456) [reply packages/wdio-appium-service/tests/launcher.test.ts]: @SrinivasanTarget @christian-bromann thanks for the explanations! I have refactored the test.

**Review by christian-bromann — COMMENTED** [review]
Thanks, the code looks good now.

Can you please help me understand how filtering out the error message helps resolve the issue? I think it makes sense adding a comment to the if statement to inform developers why we are not logging in this case.

**Review by christian-bromann — APPROVED** [review]
LGTM 👍 

Thanks so much!

**Pull request merged**